### ISSUE TITLE
Misc. bug: Could save ~3 GB VRAM in `graph_reserve` when caller doesn't need logits (big-vocab models at large ub)

### ISSUE THREAD

**Issue by drauh** [ggml-org/llama.cpp#23527]
### Name and Version

version: 9272 (1d7ab2b94)
built with GNU 13.3.0 for Linux x86_64

### Operating systems

Linux

### Which llama.cpp modules do you know to be affected?

llama-server, libllama (core library)

### Command line

```shell
./build/bin/llama-server -m /path/to/Qwen3.6-35B-A3B-UD-Q4_K_XL.gguf \
    --cpu-moe -fa on -c 16384 -b 4096 -ub 4096 \
    -ctk q8_0 -ctv q8_0 --kv-unified -np 1
```

### Problem description & steps to reproduce

`llama_context::reserve` calls `graph_reserve(n_tokens, n_seqs, n_tokens, ...)`. The third argument is `n_outputs` : the number of token rows that go through `lm_head` to produce logits, controlled per-batch by the caller via `batch.logits[i]`. Passing `n_outputs = n_tokens` here sizes the buffer for the worst case where every token in the ubatch produces a logits row.

In chat / completion workloads the caller only flags one token per sequence in `batch.logits[]`, so the executed graph runs at `n_outputs = n_seqs`, often 1. #23433 (@am17an) made the runtime cheap via `inp_out_ids`, but the reservation still sizes for the worst case. On big-vocab models this is significant.

A/B with and without a local hack capping `n_outputs` to `n_seqs` in `graph_reserve`, on Qwen3.6-35B-A3B-UD-Q4_K_XL (n_vocab=211072), RTX 3080 10 GB. Peak sampled at 10 Hz during PP of a 10001-token prompt:

| run        | idle MiB | peak MiB | PP throughput |
|------------|---------:|---------:|--------------:|
| worst case |     6597 |     6671 |  2737.65 t/s  |
| capped     |     3637 |     3711 |  2761.82 t/s  |

2960 MiB freed at idle and peak, PP within noise.

Capping at `n_seqs` only works for callers that do not want per-token logits. Callers like `examples/perplexity` that legitimately request many logits per call would break. The `llama_decode` API lets `batch.logits[i]` be set on any subset, so `graph_reserve` cannot prove `n_outputs` will be small without an explicit signal from the caller. Such a signal would also need to cover MTP / speculative-decoding verify, where the target context produces `n_max+1` logits per sequence rather than just 1.

Related:

- #23433 — runtime counterpart, `inp_out_ids` skip on `lm_head`
- #23244 — OOM with MTP on ROCm at ub=4096 (possibly same root cause)

**Comment by JohannesGaessler** [comment]
Based on my understanding of the code this sounds correct. The question is what the worst-case graphs need to cover. If you create a `llama_context` with memory for pp without logits you can at a later time still request logits, triggering a re-allocation and a potential OOM. So either we accept that requesting logits can result in a sudden increase in memory use or we require an explicit opt-in by the user and continue allocating the worst-case graphs eagerly.

**Comment by iOptimizeThings** [comment]
Hey i have a working fix for this. Tested it locally and server went from ~3.2 GB of extra VRAM at startup down to ~1.2 GB, same flags as the issue.

The fix adds a `logits_all` flag to the context params. When the server sets it to false, the reserve only sizes the buffer based on the active sequences instead of the full batch. Everything else stays worst-case by default so nothing breaks.

I already have 2 open PRs right now is it ok if I open a 3rd for this, or would you prefer I wait for one to clear first?

**Comment by am17an** [comment]
The actual compute buffer is usage at higher context is not caused by this outputs reservation. It does not provide savings at full ctx (the above instance is at 16834 context which is quite low), the bulk of the compute buffer is taken up by the kq mask, I've put a PR here to fix that https://github.com/ggml-org/llama.cpp/pull/23764

**Comment by w2jmoe** [comment]
Really interesting catch 👀

This feels like a classic “worst-case reservation vs real execution path” mismatch.

Most chat/inference workloads only need logits for a tiny subset of tokens, but the allocator still reserves as if every token flows through lm_head. On large-vocab models that becomes extremely expensive very quickly.

The ~3 GB VRAM difference here is honestly huge for 10 GB cards.

One thing that feels especially important in this report is the distinction between:

- execution-time optimization (#23433 already improved this)
- reservation-time optimization (still worst-case)

A lot of systems optimize compute but still over-reserve memory because the planner lacks intent-level information from the caller.

The idea of exposing explicit output-intent signals from the caller side seems very reasonable here, especially for future speculative / MTP flows.

Really nice investigation 👍

**Comment by drauh** [comment]
Retested at higher context. Setup: Qwen3.6-35B-A3B-UD-Q4_K_XL, `-c 150000 -b 4096 -ub 4096 -fa on -ctk q8_0 -ctv q8_0 --kv-unified --cpu-moe -np 1`, RTX 3080 10 GB. Base llama.cpp `445b7cef6` (current master), PR #23764 at `2f5190ca`. Peak sampled at 10 Hz over a 145041-token PP.

| run | CUDA0 compute | CUDA_Host compute | idle VRAM | peak VRAM |
|---|---:|---:|---:|---:|
| baseline | 3976 MiB | 2408 MiB | 7981 MiB | 8275 MiB |
| `n_outputs` cap only | 3892 MiB | 2408 MiB | 7897 MiB | 8195 MiB |
| #23764 only | 3976 MiB | 1236 MiB | 7979 MiB | 8277 MiB |
| `n_outputs` cap + #23764 | **1982 MiB** | 1236 MiB | **5985 MiB** | **6279 MiB** |

You're right that it doesn't carry the savings to higher ctx on its own, my bad for the misleading analysis...

#23764 on its own saves ~1.2 GiB, but on the `CUDA_Host` buffer it seems, not on `CUDA0`.

Combined, CUDA0 compute drops from 3976 → 1982 MiB (~2 GiB). The `lm_head` output (~3.3 GiB at ub=4096) and the f16 cast tensor (~1.2 GiB at ctx=150k) appear to be the two peaks the compute buffer is sized against; removing either alone leaves the other as the binding constraint, only removing both seems to shrinks the buffer.

**Comment by am17an** [comment]
Yes because you have a very large ub size. Default is 512.

**Comment by am17an** [comment]
Should be fixed via #23861

**Comment by drauh** [comment]
I confirm that with #23861 + #23764 I'm getting the ~2 GiB memory requirement reduction.

Thank you for implementing those two fixes. I’m now able to reach the full context size of Qwen3.6 (262k), with even some headroom left for more experts on the GPU or MTP. Amazing!

This large ub is what gives me the best PP/s, according to the tuning I did with llama-bench, so I thought it was a standard configuration.

**Comment by drauh** [comment]
Both PR are merged and together they are solving this issue.
